File: examples/vslice/src/main/scala/csw/examples/vslice/assembly/Algorithms.scala
package csw.examples.vslice.assembly

import com.typesafe.scalalogging.slf4j.LazyLogging
import csw.examples.vslice.assembly.FollowActor.CalculationConfig
import csw.examples.vslice.assembly.TromboneControl.TromboneControlConfig
import csw.util.config.DoubleItem
import csw.util.config.UnitsOfMeasure.millimeters

/**
  * TMT Source Code: 9/4/16.
  */
object Algorithms extends LazyLogging {

  /**
    * Arbitrary check of the zenith angle to be within bounds
    *
    * @param zenithAngle DoubleItem that contains zenith angle
    * @return true if valid else false
    */
  def verifyZenithAngle(zenithAngle: DoubleItem): Boolean = zenithAngle.head < 90.0 && zenithAngle.head >= 0.0

  /**
    * Checking the input focus error against fake limits of +/- 20
    *
    * @param focusError DoubleItem that contains focusError
    * @return true if valid else false
    */
  def verifyFocusError(focusError: DoubleItem): Boolean = focusError.head >= -20.0 && focusError.head <= 20.0

  // From Dan K's email,
  // elevation(t+1) =  cos(theta(t))*(elevation(t)/cos(theta(t)) + g*focusRangeDistanceError(t))
  // For this example, layer elevation is the initial elevation + a factor * cosine of zenith angle designed to be 90 to 100 km
  def naLayerElevation(calculationConfig: CalculationConfig, initialElevation: Double, zenithAngle: Double): Double =
    Math.cos(Math.toRadians(zenithAngle)) * calculationConfig.zenithFactor + initialElevation

  // The focus to range distance is  the gain * focus value/4 where focus is pinned to be between +/- 20
  def focusToRangeDistance(calculationConfig: CalculationConfig, focusError: Double): Double = {
    // Limit the focus Error
    val pinnedFocusValue = Math.max(calculationConfig.lowerFocusLimit, Math.min(calculationConfig.upperFocusLimit, focusError))
    calculationConfig.focusErrorGain * pinnedFocusValue / 4.0
  }

  /**
    * NALayerRange = zfactor*zenithAngle + focusError
    *
    * zfactor is read from configuration
    *
    * @param stagePosition is the value of the stage position in millimeters (currently the total NA elevation)
    * @return DoubleItem with key naTrombonePosition and units of enc
    */
  def rangeDistanceTransform(controlConfig: TromboneControlConfig, stagePosition: DoubleItem): Int = {
    if (stagePosition.units != millimeters) {
      // Flag an error but don't do anything else
      logger.error(s"rangeDistanceTransform: stage position units should be millimeters but are: ${stagePosition.units}")
    }
    // Scale value to be between 200 and 1000 encoder
    val encoderValue = (controlConfig.positionScale * (stagePosition.head - controlConfig.minElevation) + controlConfig.minElevationEncoder).toInt
    val pinnedEncValue = Math.max(controlConfig.minEncoderLimit, Math.min(controlConfig.maxEncoderLimit, encoderValue))
    pinnedEncValue
  }

}

File: examples/vsliceJava/src/main/java/csw/examples/vsliceJava/assembly/Algorithms.java
package csw.examples.vsliceJava.assembly;

import akka.japi.Pair;
import csw.util.config.DoubleItem;
import csw.examples.vsliceJava.assembly.AssemblyContext.*;

import static javacsw.util.config.JItems.jvalue;

/**
 * This object contains functions that implement this test version of the trombone assembly algorithms.
 * <p>
 * These algorithms are representative in spirit, but not technically correct. The actual code will be
 * a closed loop and continually recalculate the elevation. Since the correctness of the algorithms was
 * not really the focus of the vertical slice, I wanted to just get something that sort of did the
 * right things and had the same variables.
 * <p>
 * The algorithms here are implemented in an object so they can be tested independently.  I think this is
 * a good strategy.  Then they are called from the FollowActor and other places. The method names are
 * representative of their function: rangeDistanceToElevation for instance.
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class Algorithms {

  /**
   * Arbitrary check of the zenith angle to be within bounds
   *
   * @param zenithAngle DoubleItem that contains zenith angle
   * @return true if valid else false
   */
  public static boolean verifyZenithAngle(DoubleItem zenithAngle) {
    return jvalue(zenithAngle) < 90.0 && jvalue(zenithAngle) >= 0.0;
  }

  /**
   * Checking the input focus error against fake limits of +/- 20
   *
   * @param focusError DoubleItem that contains focusError
   * @return true if valid else false
   */
  public static boolean verifyFocusError(TromboneCalculationConfig calculationConfig, DoubleItem focusError) {
    return jvalue(focusError) >= calculationConfig.lowerFocusLimit && jvalue(focusError) <= calculationConfig.upperFocusLimit;
  }

  public static double zenithAngleToRangeDistance(double elevation, double zenithAngle) {
    return elevation / Math.cos(Math.toRadians(zenithAngle));
  }

  public static double rangeDistanceToElevation(double rangeDistance, double zenithAngle) {
    return Math.cos(Math.toRadians(zenithAngle)) * rangeDistance;
  }

  public static Pair<Double, Double> focusZenithAngleToElevationAndRangeDistance(TromboneCalculationConfig calculationConfig,
                                                                                 double elevation, double focusError, double zenithAngle) {
    double totalRangeDistance = focusZenithAngleToRangeDistance(calculationConfig, elevation, focusError, zenithAngle);
    double newElevation = rangeDistanceToElevation(totalRangeDistance, zenithAngle);
    return new Pair<>(totalRangeDistance, newElevation);
  }

  public static double focusZenithAngleToRangeDistance(TromboneCalculationConfig calculationConfig, double elevation, double focusError, double zenithAngle) {
    double rangeDistance1 = zenithAngleToRangeDistance(elevation, zenithAngle);
    double rangeError = focusErrorToRangeError(calculationConfig, focusError);
    return rangeDistance1 + rangeError;
  }

  // The focus to range distance is  the gain * focus value/4 where focus is pinned to be between +/- 20
  public static double focusErrorToRangeError(TromboneCalculationConfig calculationConfig, double focusError) {
    // Limit the focus Error
    double pinnedFocusValue = Math.max(calculationConfig.lowerFocusLimit, Math.min(calculationConfig.upperFocusLimit, focusError));
    return calculationConfig.focusErrorGain * pinnedFocusValue / 4.0;
  }

  /**
   * Converts the range distance is kilometers to stage position in millimeters
   * Note that in this example, the stage position is just the elevation value
   * Note that the units must be checked in the caller
   *
   * @param rangeDistance in kilometer units
   * @return stage position in millimeters
   */
  public static double rangeDistanceToStagePosition(double rangeDistance) {
    return rangeDistance;
  }

  /**
   * @param stagePosition is the value of the stage position in millimeters (currently the total NA elevation)
   * @return DoubleItem with key naTrombonePosition and units of enc
   */
  @SuppressWarnings("UnnecessaryLocalVariable")
  public static int stagePositionToEncoder(TromboneControlConfig controlConfig, double stagePosition) {
    // Scale value to be between 200 and 1000 encoder
    int encoderValue = (int) (controlConfig.positionScale * (stagePosition - controlConfig.stageZero) + controlConfig.minStageEncoder);
    int pinnedEncValue = Math.max(controlConfig.minEncoderLimit, Math.min(controlConfig.maxEncoderLimit, encoderValue));
    return pinnedEncValue;
  }

}
